src/Microsoft.Health.Fhir.Api/Features/Audit/AuditHelper.cs

```diff
@@ -84,5 +94,45 @@ private void Log(AuditAction auditAction, HttpStatusCode? statusCode, HttpContex
                     customHeaders: _auditHeaderReader.Read(httpContext));
             }
         }
+
+        private void CheckForCustomAuditHeadersInResponse(HttpContext httpContext)
+        {
+            var responseCustomHeaders = httpContext.Response.Headers.Where(x => x.Key.StartsWith(_auditConfiguration.Value.CustomAuditHeaderPrefix, StringComparison.OrdinalIgnoreCase)).ToDictionary(a => a.Key, a => a.Value.ToString());
+            if (responseCustomHeaders.Any())
+            {
+                var largeHeaders = responseCustomHeaders.Where(x => x.Value.Length > AuditConstants.MaximumLengthOfCustomHeader).ToDictionary(a => a.Key, a => a.Value.ToString());
+                if (largeHeaders.Any())
+                {
+                    throw new AuditHeaderTooLargeException(largeHeaders.First().Key, largeHeaders.First().Value.Length);
+                }
+
+                object cachedCustomHeaders;
+                var customHeaders = new Dictionary<string, string>();
+                if (httpContext.Items.TryGetValue(AuditConstants.CustomAuditHeaderKeyValue, out cachedCustomHeaders))
```

Comment:
>CustomAuditHeaderKeyValue [](http://example.com/codeflow?start=65&length=25)

This would interfere with  `auditHeaderReader.Read`, right?

Also in general is there case where you would call this method twice during request processing? If not, I would omit whole caching process, because it's not needed.